### PR TITLE
TUISwitch

### PR DESCRIPTION
**Pull request by avaidyam**
I've added a `TUISwitch` to TwUI, which replicates both the Time Machine switch, and the modern iOS switches. You can customize it using blocks.

Although it's best not used everywhere where a checkbox is used on OS X, it has its purposes (like the Time Machine prefpane).

### CODE REVIEW

File: lib/UIKit/TUISwitch.h [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+@property (nonatomic, copy) CGFloat (^knobInset)(void);
+@property (nonatomic, copy) CGFloat (^knobWidth)(void);
+@property (nonatomic, copy) CGPathRef (^knobTrackMask)(void);
+@property (nonatomic, copy) TUIViewDrawRect drawTrack;
+@property (nonatomic, copy) TUIViewDrawRect drawTrackOutline;
+@property (nonatomic, copy) TUIViewDrawRect drawKnob;
+
+- (void)setOn:(BOOL)on animated:(BOOL)animated;
+
+- (void)configureModernSwitch;
```
Comment (jwilling) [inline lib/UIKit/TUISwitch.h:33]: Having to make an explicit call to `-configure*` does not seem like an elegant solution. A custom initializer with an enum of switch types would be better.

File: lib/UIKit/TUISwitch.m [hunk 2]
```diff
@@ -0,0 +1,458 @@
...
+@property (nonatomic, readwrite, getter = knobIsGrippsed) BOOL knobGripped;
+@property (nonatomic, strong) NSMutableDictionary *twoFingersTouches;
+
+@end
+
+@implementation TUISwitch
+
+- (id)initWithFrame:(CGRect)frame {
+	if((self = [super initWithFrame:frame])) {
+		self.proxy = [[TUIView alloc] initWithFrame:self.bounds];
```
Comment (jwilling) [inline lib/UIKit/TUISwitch.m:41]: None of these initialization calls in the init method should be using setters. I don't see where they could cause a problem here, but for all general purposes it's recommended to be safe and use the ivars instead.

This applies to all the below as well.

File: lib/UIKit/TUISwitch.m [hunk 3]
```diff
@@ -0,0 +1,458 @@
...
+	knobRect.origin.x += event.deltaX;
+	
+	// If they dragged it out of the track's bounds, don't move it.
+	if(CGRectGetMinX(knobRect) < CGRectGetMinX(switchFrame) ||
+	   CGRectGetMaxX(knobRect) > CGRectGetMaxX(switchFrame))
+		return YES;
+	
+	// If the knob is still within bounds, interpolate it by
+	// using a linear animation curve, so any "jumps" aren't seen.
+	[TUIView animateWithDuration:0.1 animations:^{
```
Comment (jwilling) [inline lib/UIKit/TUISwitch.m:374]: Animating every time a new mouse event is received causes a perception of lag when dragging the knob around. This should not be animated.

File: lib/UIKit/TUISwitch.m [hunk 4]
```diff
@@ -0,0 +1,458 @@
...
+			[TUIView setAnimationCurve:TUIViewAnimationCurveEaseInOut];
+			self.knob.frame = knobRect;
+			self.toggle.frame = toggleRect;
+			
+			[self.knob redraw];
+			[self.outline redraw];
+			[self.toggle redraw];
+		}];
+	} else {
+		self.knob.frame = knobRect;
```
Comment (jwilling) [inline lib/UIKit/TUISwitch.m:422]: Duplicated code here. Have you considered setting the duration of the animation to zero if it shouldn't be animated?

**Comment by jwilling** [comment]
I'm going to have to make a call here and say this should not be included in TwUI. Switches really are meant for touch screens, and I see TwUI having a far greater need for a standard checkbox.

I'd encourage you to start a repo containing miscellaneous user interface elements that aren't included in the main repo. This would be a perfect fit.

**Comment by avaidyam** [comment]
I've noted that it should not be used instead of checkboxes, but, it does have its appropriate use, even on OS X. The time machine prefpane is an example - a simple on or off switch for the entire feature. 

For checkboxes, I'm working on the TUIButton class to bring that up to par with UIButton and NSButton, so for everything else that is either a YES or NO action, you would use that.

**Comment by jspahrsummers** [comment]
Apple's use of a switch in a single preference pane doesn't mean that the control is generally useful or common.

**Comment by avaidyam** [comment]
To be fair, I did know that when I designed this ;) I don't mind if it's rejected from TwUI, but I do see the use of a switch where appropriate, and if given the time, could list many different uses. If you'd like to close this, I'd be fine with it, but I'm sure there is some place for it in TwUI (multitouch trackpads, and gesture-based interfaces).

**Comment by jwilling** [comment]
Closing. Please create a repo that contains controls of this kind, as I'm sure others will find them helpful.

**Comment by avaidyam** [comment]
Alright, thanks. And to that effect, would a `TUICellControl` be useful? It'd take an `NSCell` and wrap it in a `TUIControl`, achieving near-perfect synchronism with most `NSControls`. A very basic example I have working right now is a `TUICellControl` with a `cellClass` of `NSButtonCell`. It works so-so, and with a bit of patch-up, it'd work JUST like an `NSButton`, theoretically (after testing).

**Comment by jspahrsummers** [comment]
Please create a separate issue.

**Pull request closed without merging**